src/Streamly/Internal/Data/Array/Mut/Type.hs

```diff
@@ -0,0 +1,535 @@
+snoc = snocWith (* 2)
+
+-------------------------------------------------------------------------------
+-- Random reads
+-------------------------------------------------------------------------------
+
+-- | Return the element at the specified index without checking the bounds.
+--
+-- Unsafe because it does not check the bounds of the array.
+{-# INLINE_NORMAL getIndexUnsafe #-}
+getIndexUnsafe :: MonadIO m => Array a -> Int -> m a
+getIndexUnsafe Array {..} n =
+    liftInIO
+        $ \s# ->
+              let !(I# i#) = arrStart + n
+               in readArray# arrContents# i# s#
+
+-- | /O(1)/ Lookup the element at the given index. Index starts from 0.
+--
+{-# INLINE getIndex #-}
+getIndex :: MonadIO m => Array a -> Int -> m a
+getIndex arr@(Array {..}) i =
+    if i < arrLen && i > 0
+    then getIndexUnsafe arr i
+    else invalidIndex "getIndex" i
+
+-------------------------------------------------------------------------------
+-- Subarrays
+-------------------------------------------------------------------------------
+
+-- XXX We can also get immutable slices.
+
+-- | /O(1)/ Slice an array in constant time.
+--
+-- Unsafe: The bounds of the slice are not checked.
+--
+-- /Unsafe/
+--
+-- /Pre-release/
+{-# INLINE getSliceUnsafe #-}
+getSliceUnsafe
+    :: Int -- ^ from index
+    -> Int -- ^ length of the slice
+    -> Array a
+    -> Array a
+getSliceUnsafe index len arr@(Array {..}) =
+    assert (index >= 0 && len >= 0 && index + len <= arrLen)
+        $ arr {arrStart = arrStart + index, arrLen = len}
+
+
+-- | /O(1)/ Slice an array in constant time. Throws an error if the slice
+-- extends out of the array bounds.
+--
+-- /Pre-release/
+{-# INLINE getSlice #-}
+getSlice
+    :: Int -- ^ from index
+    -> Int -- ^ length of the slice
+    -> Array a
+    -> Array a
+getSlice index len arr@(Array{..}) =
+    if index >= 0 && len >= 0 && index + len <= arrLen
+    then arr {arrStart = arrStart + index, arrLen = len}
+    else error
+             $ "getSlice: invalid slice, index "
+             ++ show index ++ " length " ++ show len
+
+-------------------------------------------------------------------------------
+-- to Lists and streams
+-------------------------------------------------------------------------------
+
+-- | Convert an 'Array' into a list.
+--
+-- /Pre-release/
+{-# INLINE toList #-}
+toList :: MonadIO m => Array a -> m [a]
+toList arr@(Array {..}) = mapM (getIndexUnsafe arr) [0 .. (arrLen - 1)]
+
+-- | Use the 'read' unfold instead.
+--
+-- @toStreamD = D.unfold read@
+--
+-- We can try this if the unfold has any performance issues.
+{-# INLINE_NORMAL toStreamD #-}
+toStreamD :: MonadIO m => Array a -> D.Stream m a
+toStreamD arr@(Array {..}) =
+    D.mapM (getIndexUnsafe arr) $ D.enumerateFromToIntegral 0 (arrLen - 1)
+
+{-# INLINE toStreamK #-}
+toStreamK :: MonadIO m => Array a -> K.Stream m a
+toStreamK arr@(Array {..}) = K.mapM (getIndexUnsafe arr) $ K.unfoldr step 0
```

Comment:
Would it be faster if we K.unfoldrM instead of using mapM on unfoldr? depends on whether mapM can fuse, I guess it doesn't for CPS streams.